api/inc/register_gateway.h

```diff
@@ -17,104 +17,210 @@
 #ifndef __UVISOR_API_REGISTER_GATEWAY_H__
 #define __UVISOR_API_REGISTER_GATEWAY_H__
 
+#include "api/inc/register_gateway_exports.h"
 #include "api/inc/uvisor_exports.h"
 #include <stdint.h>
 
-/* The following magic is used to verify that a register gateway structure. It
- * has been generated starting from the ARM Thumb/Thumb-2 invalid opcode.
- * ARM Thumb (w/o the Thumb-2 extension): 0xF7FxAxxx (32 bits)
- * The 'x's can be freely chosen (they have been chosen randomly here).
- * This requires the Thumb-2 extensions because otherwise only 16 bits opcodes
- * are accepted.*/
-#if defined(__thumb__) && defined(__thumb2__)
-#define UVISOR_REGISTER_GATEWAY_MAGIC 0xF7F3A89E
-#else
-#error "Unsupported instruction set. The ARM Thumb-2 instruction set must be supported."
-#endif /* __thumb__ && __thumb2__ */
-
-/* register gateway operations */
-/* note: do not use special characters as these numbers will be stringified */
-#define UVISOR_OP_READ(op)  (op)
-#define UVISOR_OP_WRITE(op) ((1 << 4) | (op))
-#define UVISOR_OP_NOP       0x0
-#define UVISOR_OP_AND       0x1
-#define UVISOR_OP_OR        0x2
-#define UVISOR_OP_XOR       0x3
-
-/* default mask for whole register operatins */
-#define __UVISOR_OP_DEFAULT_MASK 0x0
-
-/* register gateway metadata */
-#if defined(__CC_ARM)
-
-/* TODO/FIXME */
-
-#elif defined(__GNUC__)
-
-/* 1 argument: simple read, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA1(src_box, addr) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
+/** Get the offset of a struct member.
+ * @internal
+ */
+#define OFFSETOF(type, member) ((uint32_t) (&(((type *)(0))->member)))
```

Comment:
@AlessandroA we needs to namespace it anyway for uVisor, right - UVISOR_OFFSET_OF ?